packages/cli/src/commands/validator/missed-blocks.ts

```diff
@@ -0,0 +1,149 @@
+    ...BaseCommand.flagsWithoutLocalAddresses(),
+    genesis: Flags.path({
+      description:
+        'path to the genesis block to get the initial validator set, uses contracts if not provided.',
+    }),
+    'at-block': flags.integer({
+      description: 'latest block to examine for signer activity',
+    }),
+    lookback: flags.integer({
+      description: 'how many blocks to look back for signer activity',
+      default: 120,
+    }),
+  }
+
+  static examples = [
+    'downtime --genesis ./genesis.json',
+    'downtime',
+    'downtime --at-block 100000 --genesis ./genesis.json',
+    'downtime --lookback 500 --genesis ./genesis.json',
+  ]
+
+  async run() {
+    const res = this.parse(ValidatorSignedBlocks)
+
+    const latest = res.flags['at-block']
+      ? res.flags['at-block'] + 1
+      : (await this.web3.eth.getBlock('latest')).number
+
+    let electionCache
+    let validators, accounts
+    let haveContracts
+    let electedSigners: string[]
+    const validatorToGroupName = new Map<Address, string>()
+    // Try to get contracts, if not try to fall back to genesis block
+    try {
+      const election = await this.kit.contracts.getElection()
+      accounts = await this.kit.contracts.getAccounts()
+      validators = await this.kit.contracts.getValidators()
+      const vgroups = await validators.getRegisteredValidatorGroups()
+      for (const group of vgroups) {
+        for (const member of group.members) {
+          validatorToGroupName.set(member, group.name || '')
+        }
+      }
+      const epochSize = await validators.getEpochSize()
+      electionCache = new ElectionResultsCache(election, epochSize.toNumber())
+      electedSigners = await electionCache.electedSigners(latest)
+      haveContracts = true
+    } catch {
+      if (res.flags.genesis) {
+        const genesisJson = JSON.parse(readFileSync(res.flags.genesis, 'utf-8'))
+        electedSigners = parseBlockExtraData(genesisJson.extraData).addedValidators
+      } else {
+        this.error('If contracts are not deployed, must provide the genesis block')
+      }
+      haveContracts = false
+    }
+
+    // Validators added from extra data in genesis
+    const missedBlocks = new Map<string, number[]>()
+
+    // Get blocks
+    const blocks = await concurrentMap(10, [...Array(res.flags.lookback).keys()], (i) =>
+      this.web3.eth.getBlock(latest - res.flags.lookback! + i + 1)
+    )
+    // Calculate uptime
+    for (const block of blocks) {
+      // Grab new validators on new epoch
+      if (haveContracts) {
+        // @ts-ignore
+        electedSigners = await electionCache.electedSigners(block.number)
+      }
+      for (let i = 0; i < electedSigners.length; i++) {
+        if (!missedBlocks.has(electedSigners[i])) {
+          missedBlocks.set(electedSigners[i], [])
+        }
+        const bitmap = parseBlockExtraData(block.extraData).parentAggregatedSeal.bitmap
+        if (!bitIsSet(bitmap, i)) {
+          // @ts-ignore
```

Comment:
why the ignore?

Reply:
Typescript can't realize that I check if the element is present, and set a value for it if not just above. None of the common workarounds here work without making the code much more complex.